File: examples/object-detection/multistream-yolox-yolov8-object-detection-optiview/cpp/utils/sample_utils.cpp
#include "sample_utils_api.cpp"

#include "neat/session.h"
#include "support/object_detection/obj_detection_utils.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace multistream_yolox_yolov8_optiview {
namespace {

std::string sample_payload_tag_upper(const simaai::neat::Sample& sample) {
  std::string tag = sample.payload_tag.empty() ? sample.format : sample.payload_tag;
  tag = objdet::upper_ascii_copy(tag);
  if (!tag.empty()) {
    return tag;
  }
  for (const auto& field : sample.fields) {
    tag = sample_payload_tag_upper(field);
    if (!tag.empty()) {
      return tag;
    }
  }
  return {};
}

std::string class_label(const std::vector<std::string>& class_labels, int class_id) {
  if (class_id >= 0 && class_id < static_cast<int>(class_labels.size())) {
    return class_labels[static_cast<std::size_t>(class_id)];
  }
  return std::to_string(class_id);
}

const simaai::neat::Tensor* first_tensor_impl(const simaai::neat::Sample& sample) {
  if (sample.kind == simaai::neat::SampleKind::Tensor && sample.tensor.has_value()) {
    return &sample.tensor.value();
  }
  for (const auto& field : sample.fields) {
    if (const auto* tensor = first_tensor_impl(field); tensor != nullptr) {
      return tensor;
    }
  }
  return nullptr;
}

} // namespace

std::string to_string(DetectorOutputKind kind) {
  switch (kind) {
  case DetectorOutputKind::BBox:
    return "BBOX";
  case DetectorOutputKind::DetessDequant:
    return "DETESSDEQUANT";
  }
  return "UNKNOWN";
}

std::vector<std::uint8_t> extract_bbox_payload(const simaai::neat::Sample& sample) {
  std::vector<std::uint8_t> payload;
  std::string error;
  if (!objdet::extract_bbox_payload(sample, payload, error)) {
    return {};
  }
  return payload;
}

std::vector<Detection> parse_bbox_payload(const std::vector<std::uint8_t>& payload, int img_w,
                                          int img_h) {
  const auto parsed = objdet::parse_boxes_lenient(payload, img_w, img_h, 0);
  std::vector<Detection> out;
  out.reserve(parsed.size());
  for (const auto& box : parsed) {
    out.push_back(Detection{
        box.x1,
        box.y1,
        box.x2,
        box.y2,
        box.score,
        box.class_id,
    });
  }
  return out;
}

DetectorOutputKind require_detector_output_kind(ModelFamily family,
                                                const simaai::neat::Sample& sample) {
  std::string expected;
  DetectorOutputKind kind = DetectorOutputKind::BBox;
  switch (family) {
  case ModelFamily::YoloV8:
    expected = "BBOX";
    kind = DetectorOutputKind::BBox;
    break;
  case ModelFamily::YoloX:
    expected = "DETESSDEQUANT";
    kind = DetectorOutputKind::DetessDequant;
    break;
  case ModelFamily::Auto:
    throw std::invalid_argument("unsupported model family");
  }

  const std::string actual = sample_payload_tag_upper(sample);
  if (actual != expected) {
    throw std::runtime_error("unsupported detector output: family=" + to_string(family) +
                             " expected=" + expected + " actual=" +
                             (actual.empty() ? std::string("<empty>") : actual));
  }
  return kind;
}

std::vector<Detection> detections_from_detector_sample(ModelFamily family,
                                                       const simaai::neat::Sample& sample,
                                                       int img_w, int img_h) {
  const DetectorOutputKind kind = require_detector_output_kind(family, sample);
  const auto payload = extract_bbox_payload(sample);
  if (!payload.empty()) {
    return parse_bbox_payload(payload, img_w, img_h);
  }
  if (kind == DetectorOutputKind::DetessDequant) {
    throw std::runtime_error(
        "YOLOX detector output did not expose a BBOX payload; "
        "DetessDequant output parsing could not be finalized for this model pack");
  }
  return {};
}

std::string optiview_frame_id(const simaai::neat::Sample& sample, int fallback_frame_index) {
  if (sample.frame_id >= 0) {
    return std::to_string(sample.frame_id);
  }
  return std::to_string(fallback_frame_index);
}

std::int64_t optiview_timestamp_ms(double publish_time_s, double offset_ms) {
  return static_cast<std::int64_t>(std::llround(publish_time_s * 1000.0 + offset_ms));
}

OptiViewDetectionPayload build_optiview_detection_payload(const std::vector<Detection>& detections,
                                                          int img_w, int img_h,
                                                          const std::vector<std::string>& class_labels) {
  OptiViewDetectionPayload payload;
  payload.objects.reserve(detections.size());
  for (const auto& det : detections) {
    const int x1 = std::max(0, std::min(img_w, static_cast<int>(det.x1)));
    const int y1 = std::max(0, std::min(img_h, static_cast<int>(det.y1)));
    const int x2 = std::max(0, std::min(img_w, static_cast<int>(det.x2)));
    const int y2 = std::max(0, std::min(img_h, static_cast<int>(det.y2)));
    if (x2 <= x1 || y2 <= y1) {
      continue;
    }

    sima_examples::OptiViewObject object;
    object.x = x1;
    object.y = y1;
    object.w = std::max(0, x2 - x1);
    object.h = std::max(0, y2 - y1);
    object.score = det.score;
    object.class_id = det.class_id;
    payload.objects.push_back(object);
  }
  // OptiView resolves object labels by indexing this full label table with object.class_id.
  payload.labels = class_labels;
  return payload;
}

cv::Mat tensor_rgb_from_sample(const simaai::neat::Sample& sample) {
  const auto* tensor = first_tensor_impl(sample);
  if (tensor == nullptr) {
    throw std::runtime_error("no tensor payload found in decoded RTSP sample");
  }
  return tensor->to_cv_mat_copy(simaai::neat::ImageSpec::PixelFormat::RGB);
}

} // namespace multistream_yolox_yolov8_optiview

File: examples/object-detection/multistream-yolox-yolov8-object-detection-optiview/python/utils/sample_utils.py
"""Sample helpers for family-specific detector outputs."""

from __future__ import annotations

import struct
from typing import Any

from .model_family import YOLOX_NOT_SUPPORTED_MESSAGE


def class_label(class_labels: list[str], class_id: int) -> str:
    if 0 <= int(class_id) < len(class_labels):
        return class_labels[int(class_id)]
    return str(int(class_id))


def iter_tensors(pyneat: Any, sample: Any):
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        yield sample.tensor
    for field in getattr(sample, "fields", []):
        yield from iter_tensors(pyneat, field)


def extract_bbox_payload(pyneat: Any, sample: Any) -> bytes | None:
    stack = [sample]
    while stack:
        current = stack.pop()
        stack.extend(reversed(list(getattr(current, "fields", []))))
        if getattr(current, "kind", None) != pyneat.SampleKind.Tensor:
            continue
        tensor = getattr(current, "tensor", None)
        if tensor is None:
            continue
        fmt = (getattr(current, "payload_tag", "") or getattr(current, "format", "") or "").upper()
        if fmt and fmt != "BBOX":
            continue
        try:
            payload = tensor.copy_payload_bytes()
        except Exception:
            continue
        if payload:
            return payload
    return None


def parse_bbox_payload(payload: bytes | None, img_w: int, img_h: int) -> list[dict]:
    if not payload or len(payload) < 4:
        return []
    count = min(struct.unpack_from("<I", payload, 0)[0], (len(payload) - 4) // 24)
    boxes: list[dict] = []
    offset = 4
    for _ in range(count):
        x, y, w, h, score, class_id = struct.unpack_from("<iiiifi", payload, offset)
        offset += 24
        x1 = max(0.0, min(float(img_w), float(x)))
        y1 = max(0.0, min(float(img_h), float(y)))
        x2 = max(0.0, min(float(img_w), float(x + w)))
        y2 = max(0.0, min(float(img_h), float(y + h)))
        if x2 <= x1 or y2 <= y1:
            continue
        boxes.append(
            {
                "x1": x1,
                "y1": y1,
                "x2": x2,
                "y2": y2,
                "score": float(score),
                "class_id": int(class_id),
            }
        )
    return boxes


def sample_payload_tag(pyneat: Any, sample: Any) -> str:
    if sample is None:
        return ""
    tag = (getattr(sample, "payload_tag", "") or getattr(sample, "format", "") or "").upper()
    if tag:
        return tag
    for field in getattr(sample, "fields", []):
        child = sample_payload_tag(pyneat, field)
        if child:
            return child
    return ""


def optiview_frame_id(sample: Any, fallback_frame_index: int) -> str:
    frame_id = getattr(sample, "frame_id", -1)
    try:
        if int(frame_id) >= 0:
            return str(int(frame_id))
    except Exception:
        pass
    return str(int(fallback_frame_index))


def optiview_timestamp_ms(publish_time_s: float, offset_ms: float) -> int:
    return int(round(float(publish_time_s) * 1000.0 + float(offset_ms)))


def require_detector_output_kind(pyneat: Any, family: str, sample: Any) -> str:
    lowered = str(family).strip().lower()
    if lowered == "yolov8":
        expected = "BBOX"
    elif lowered == "yolox":
        raise ValueError(YOLOX_NOT_SUPPORTED_MESSAGE)
    else:
        raise ValueError(f"unsupported model family: {family}")

    actual = sample_payload_tag(pyneat, sample)
    if actual != expected:
        raise RuntimeError(
            f"unsupported detector output: family={lowered} expected={expected} actual={actual or '<empty>'}"
        )
    return expected


def detections_from_detector_sample(
    pyneat: Any,
    family: str,
    sample: Any,
    img_w: int,
    img_h: int,
) -> list[dict]:
    kind = require_detector_output_kind(pyneat, family, sample)
    payload = extract_bbox_payload(pyneat, sample)
    if payload:
        return parse_bbox_payload(payload, img_w, img_h)
    if kind != "BBOX":
        raise RuntimeError(f"unsupported detector output kind: {kind}")
    return []


def make_optiview_detection_payload(
    pyneat: Any,
    detections: list[dict],
    *,
    img_w: int,
    img_h: int,
    class_labels: list[str],
) -> tuple[list[Any], list[str]]:
    objects: list[Any] = []
    for det in detections:
        x1 = max(0.0, min(float(img_w), float(det.get("x1", 0.0))))
        y1 = max(0.0, min(float(img_h), float(det.get("y1", 0.0))))
        x2 = max(0.0, min(float(img_w), float(det.get("x2", 0.0))))
        y2 = max(0.0, min(float(img_h), float(det.get("y2", 0.0))))
        if x2 <= x1 or y2 <= y1:
            continue

        obj = pyneat.OptiViewObject()
        obj.x = int(x1)
        obj.y = int(y1)
        obj.w = int(max(0.0, x2 - x1))
        obj.h = int(max(0.0, y2 - y1))
        obj.score = float(det.get("score", 0.0))
        obj.class_id = int(det.get("class_id", -1))
        objects.append(obj)
    # OptiView resolves object labels by indexing this full label table with object.class_id.
    return objects, list(class_labels)


def first_tensor(pyneat: Any, sample: Any) -> Any:
    if sample is None:
        return None
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        return sample.tensor
    for tensor in iter_tensors(pyneat, sample):
        return tensor
    return None


def tensor_rgb_from_sample(runtime, sample: Any):
    tensor = first_tensor(runtime.pyneat, sample)
    if tensor is None:
        raise RuntimeError("no tensor payload found in decoded RTSP sample")
    arr = runtime.np.asarray(tensor.to_numpy(copy=True))
    if arr.ndim == 4 and arr.shape[0] == 1:
        arr = arr[0]
    if arr.ndim != 3:
        raise RuntimeError(f"unexpected decoded tensor shape: {arr.shape}")
    if arr.dtype != runtime.np.uint8:
        arr = runtime.np.clip(arr, 0, 255).astype(runtime.np.uint8)
    return runtime.np.ascontiguousarray(arr)
